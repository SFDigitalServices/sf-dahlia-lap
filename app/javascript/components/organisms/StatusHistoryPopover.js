import React, { useState } from 'react'

import classNames from 'classnames'
import PropTypes from 'prop-types'

import apiService from 'apiService'
import { COLORS } from 'components/atoms/colors'
import StyledIcon from 'components/atoms/StyledIcon'
import StatusItems from 'components/molecules/lease_up_sidebar/StatusItems'
import Popover from 'components/molecules/Popover'

import Loading from '../molecules/Loading'

const MAX_UPDATES_TO_SHOW_DEFAULT = 4

const StatusHistoryPopover = ({ applicationId, customButtonClasses = null }) => {
  const [showingAllStatuses, setShowingAllStatuses] = useState(false)
  const [state, setState] = useState({ statusItems: [], loading: false })

  const onShowHideStatusesToggled = () => setShowingAllStatuses(!showingAllStatuses)

  const fetchStatusHistory = () => {
    setState({ statusItems: state.statusItems, loading: true })
    apiService
      .getFieldUpdateComments(applicationId)
      .then((items) => setState({ statusItems: items, loading: false }))
      .catch(() => setState({ statusItems: state.statusItems, loading: false }))
  }

  const onClick = () => {
    state.statusItems.length === 0 && fetchStatusHistory()
  }

  const numberOfStatusesToDisplay = showingAllStatuses
    ? state.statusItems.length
    : MAX_UPDATES_TO_SHOW_DEFAULT

  const showShowMoreToggler = state.statusItems.length > MAX_UPDATES_TO_SHOW_DEFAULT
  return (
    <Popover
      buttonElement={({ onClick, ref }) => (
        <button
          ref={ref}
          onClick={onClick}
          className={classNames(customButtonClasses, 'button-icon-only')}
        >
          <StyledIcon icon='list-unordered' customSizeRem='1.2rem' customFill={COLORS.steel} />
        </button>
      )}
      onButtonClick={onClick}
    >
      <Loading isLoading={state.loading}>
        <StatusItems
          statusItems={state.statusItems}
          limit={numberOfStatusesToDisplay}
          height={showShowMoreToggler ? '20rem' : null}
        />
        {showShowMoreToggler && (
          <button
            className={classNames('button-link', 't-tiny')}
            type='button'
            onClick={onShowHideStatusesToggled}
          >
            {showingAllStatuses ? 'Show only recent status updates' : 'Show all status updates'}
          </button>
        )}
      </Loading>
    </Popover>
  )
}
StatusHistoryPopover.propTypes = {
  applicationId: PropTypes.string.isRequired,
  customButtonClasses: PropTypes.string
}

export default StatusHistoryPopover
