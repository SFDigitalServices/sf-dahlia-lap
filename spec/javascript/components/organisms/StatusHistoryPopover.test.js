import React from 'react'

import { mount } from 'enzyme'
import { act } from 'react-dom/test-utils'

import StatusHistoryPopover from 'components/organisms/StatusHistoryPopover'

import { mockManyStatusItems } from '../../mocks/statusItemMock'

const mockGetFieldUpdateComments = jest.fn()
const APP_ID = 'app_id'
const APP_ID_MANY_ITEMS = 'app_id_many_items'

jest.mock('apiService', () => {
  return {
    getFieldUpdateComments: async (applicationId) => {
      mockGetFieldUpdateComments(applicationId)
      // Can't use const for id here because mock cannot use out of scope variables.
      return applicationId === 'app_id_many_items' ? mockManyStatusItems(5) : mockManyStatusItems(2)
    }
  }
})

const tick = () => {
  return new Promise((resolve) => {
    setTimeout(resolve, 0)
  })
}

const getWrapper = (appId) => mount(<StatusHistoryPopover applicationId={appId} />)

let wrapper
describe('StatusHistoryPopover', () => {
  describe('when there are fewer than 4 status items', () => {
    beforeEach(async () => {
      wrapper = getWrapper(APP_ID)
      act(() => {
        wrapper.find('button').simulate('click')
      })
      await act(async () => await tick)
      wrapper.update()
      return wrapper
    })
    test('it matches snapshot when open', async () => {
      expect(wrapper).toMatchSnapshot()
    })
    test("it loads status history data when data isn't present", async () => {
      expect(mockGetFieldUpdateComments).toHaveBeenCalledTimes(1)
    })
    test('it only fetches status history on first click', async () => {
      // Click again and make sure history items aren't fetched again
      act(() => {
        wrapper.find('button').simulate('click')
      })
      await act(async () => await tick)
      wrapper.update()
      expect(mockGetFieldUpdateComments).toHaveBeenCalledTimes(1)
    })

    test('does not render the show/hide status toggler', async () => {
      expect(wrapper.exists('.button-link')).not.toEqual(true)
    })

    test('does not set a fixed height for status items', () => {
      expect(wrapper.find('StatusItems').prop('height')).toBeNull()
    })
  })
  describe('when there are more than 4 status items', () => {
    beforeEach(async () => {
      wrapper = getWrapper(APP_ID_MANY_ITEMS)
      act(() => {
        wrapper.find('button').simulate('click')
      })
      await act(async () => await tick)
      wrapper.update()
      return wrapper
    })
    test('renders a show/hide link that shows and hides status items', async () => {
      expect(wrapper.exists('.button-link')).toEqual(true)
      expect(wrapper.find('.button-link').text()).toEqual('Show all status updates')
      expect(wrapper.find('StatusItem')).toHaveLength(4)
      wrapper.find('.button-link').simulate('click')
      expect(wrapper.find('StatusItem')).toHaveLength(5)
      expect(wrapper.find('.button-link').text()).toEqual('Show only recent status updates')
    })

    test('sets a fixed height for the status items component', async () => {
      expect(wrapper.find('StatusItems').prop('height')).toEqual('20rem')
    })
  })
})
