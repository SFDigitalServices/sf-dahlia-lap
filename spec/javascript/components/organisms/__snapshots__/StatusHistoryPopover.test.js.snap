// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusHistoryPopover when there are fewer than 4 status items it matches snapshot when open 1`] = `
<StatusHistoryPopover
  applicationId="app_id"
>
  <Popover
    buttonElement={[Function]}
    onButtonClick={[Function]}
  >
    <button
      className="button-icon-only"
      onClick={[Function]}
    >
      <StyledIcon
        customFill="#767676"
        customSizeRem="1.2rem"
        icon="list-unordered"
      >
        <svg
          style={
            Object {
              "fill": "#767676",
              "height": "1.2rem",
              "width": "1.2rem",
            }
          }
        >
          <use
            style={
              Object {
                "fill": "#767676",
              }
            }
            xlinkHref="#i-list-unordered"
          />
        </svg>
      </StyledIcon>
    </button>
    <div
      className="popper-container"
      data-popper-escaped={true}
      data-popper-placement="bottom"
      data-popper-reference-hidden={true}
      style={
        Object {
          "bottom": "auto",
          "left": "0",
          "position": "absolute",
          "right": "auto",
          "top": "0",
          "transform": "translate(0px, 32px)",
        }
      }
    >
      <div
        id="arrow"
        style={
          Object {
            "left": "0",
            "position": "absolute",
            "top": "",
            "transform": "translate(0px, 0px)",
          }
        }
      />
      <Loading
        isLoading={false}
      >
        <div
          className="loading-panel"
        >
          <StatusItems
            height={null}
            limit={4}
            statusItems={
              Array [
                Object {
                  "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
                  "status": "Approved",
                  "substatus": "Approval letter sent",
                  "timestamp": 1598385600,
                },
                Object {
                  "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
                  "status": "Approved",
                  "substatus": "Approval letter sent",
                  "timestamp": 1598385600,
                },
              ]
            }
          >
            <div
              className="status-items"
              style={Object {}}
            >
              <StatusItem
                key="0"
                statusItem={
                  Object {
                    "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
                    "status": "Approved",
                    "substatus": "Approval letter sent",
                    "timestamp": 1598385600,
                  }
                }
              >
                <div
                  className="status-item"
                >
                  <div
                    className="status-item-header"
                  >
                    <StatusPill
                      status="Approved"
                    >
                      <div
                        className="status-pill is-approved"
                      >
                        Approved
                      </div>
                    </StatusPill>
                    <StatusDate
                      timestamp={1598385600}
                    >
                      <div
                        className="status-item-date"
                      >
                        Aug 25, 2020
                      </div>
                    </StatusDate>
                  </div>
                  <div
                    className="item-substatus-text"
                  >
                    Approval letter sent
                  </div>
                  <div
                    className="status-item-text"
                  >
                    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.
                  </div>
                </div>
              </StatusItem>
              <StatusItem
                key="1"
                statusItem={
                  Object {
                    "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.",
                    "status": "Approved",
                    "substatus": "Approval letter sent",
                    "timestamp": 1598385600,
                  }
                }
              >
                <div
                  className="status-item"
                >
                  <div
                    className="status-item-header"
                  >
                    <StatusPill
                      status="Approved"
                    >
                      <div
                        className="status-pill is-approved"
                      >
                        Approved
                      </div>
                    </StatusPill>
                    <StatusDate
                      timestamp={1598385600}
                    >
                      <div
                        className="status-item-date"
                      >
                        Aug 25, 2020
                      </div>
                    </StatusDate>
                  </div>
                  <div
                    className="item-substatus-text"
                  >
                    Approval letter sent
                  </div>
                  <div
                    className="status-item-text"
                  >
                    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.
                  </div>
                </div>
              </StatusItem>
            </div>
          </StatusItems>
        </div>
      </Loading>
    </div>
  </Popover>
</StatusHistoryPopover>
`;
